x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml

```diff
@@ -91,6 +92,25 @@ setup:
                     boost: 3
               lang: "mustache"
 
+  - do:
+      search_application.put:
```

Comment:
Rather than throw on `search` and `render_template` I think the better option would be to throw here on the `put` command - Do the validation when we store the template, and never store an invalid template. WDYT?

Reply:
I agree that would be optimal, but it would require larger and (probably) breaking changes. For example, throwing when creating/updating the search application would require default values for all params so we have values to use for a test render. And we may need to require param validation specs to ensure that param data types and values passed to `_render_query` or `_search` are valid.

Maybe we create a follow-up issue to investigate this?

Reply:
I would prefer to fix this the right way from the start. Perhaps we can time box an investigation on how to fix this as part of your issue and discuss further? 

I don't understand why this would require breaking changes, but please note that search applications is still a beta feature, so if you identify a breaking change it's not something we love to do but it's something that's on the table if it's the right thing to do.

Reply:
The issue I see with moving all validation to search application create/update time is that it puts a heavy burden on the user to provide test values for all params.

The only way I see to validate a template at search application create/update time is to perform one or more test renders of the template, which requires test values. Additionally, these test values must be representative of edge cases that could break the template. And since it's easy to have multiple edge cases, we would need to support collecting multiple sets of test values from the user. And even then, the test render(s) would only be as good as the test values provided by the user, so could we really trust them? We would still need to gracefully handle a template rendering error in `_render_query` or `_search` (to account for when the user does not sufficiently test an edge case), and thus the code introduced in this PR would still be necessary.

The key theoretical point that complicates your request is that a template's validity is determined by a combination of the template itself and the param values used to render the template. We only have the template at search application create/update time and as described above, we would need to rely on the user to provide value(s).  Relying on user input for such an important validation step is bound to backfire on us.

Reply:
OK, good point, I'm on board with keeping the defensive checks in the search and render metadata calls. 

However, I'd still like to see what we can do about solving this at creation time even if it's not a complete solution.

Reply:
I agree that we can keep this as is and think of a more complete solution for preventing this to happen. Provide test cases as part of creating or updating a template might make sense, let's discuss 👍